### PR TITLE
JAPI-303

### PR DESCRIPTION
**Pull request by jpmcmu**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [X] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).

## Checklist:
- [X] I have created a corresponding JIRA ticket for this submission
- [X] My code follows the code style of this project.
  - [ ] I have applied the Eclipse code-format template provided.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [X] I have read the HPCC Systems CONTRIBUTORS document (https://github.com/hpcc-systems/HPCC-Platform/wiki/Guide-for-contributors).
- [X] The change has been fully tested:
  - [ ] I have performed unit tests to cover my changes.
  - [ ] I have performed system test and covered possible regressions and side effects.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.

### CODE REVIEW

**Comment by jpmcmu** [comment]
@rpastrana still WIP but please review

**Comment by jpmcmu** [comment]
@rpastrana please review

File: commons-hpcc/src/main/java/org/hpccsystems/commons/benchmarking/BenchmarkResult.java [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+            JSONObject param = parameters.get(i).toJson();
+            jsonParams.put(param);
+        }
+
+        test.put("parameters",parameters);
+
+        // Write test results
+        JSONArray jsonMetrics = new JSONArray();
+
+        if (selectedResults == null)
```
Comment (rpastrana) [inline commons-hpcc/src/main/java/org/hpccsystems/commons/benchmarking/BenchmarkResult.java:112]: or selectedResults.length == 0 ?

File: commons-hpcc/src/main/java/org/hpccsystems/commons/benchmarking/BenchmarkResult.java [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        {
+            for (int i = 0; i < selectedResults.length; i++)
+            {
+                AveragedMetric metric = metrics.get(selectedResults[i]);
+                if (metric == null)
+                {
+                    continue;
+                }
+
+                metric.discardOutliers();
```
Comment (rpastrana) [inline commons-hpcc/src/main/java/org/hpccsystems/commons/benchmarking/BenchmarkResult.java:133]: are there any reasonable scenarios where a user wouldn't want to discard outliers as determined by the THOMPSON_TAU logic?

File: commons-hpcc/src/main/java/org/hpccsystems/commons/benchmarking/AveragedMetric.java [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+            float stdDeviation = 0;
+            float largestAbsDeviation = Float.MIN_VALUE;
+            int largestAbsDeviationIndex = 0;
+
+            for (int i = 0; i < dataPoints.size(); i++)
+            {
+                float absoluteDeviation = dataPoints.get(i).floatValue() - avg;
+                if (largestAbsDeviation < Math.abs(absoluteDeviation))
+                {
+                    largestAbsDeviation = absoluteDeviation;
```
[the remark this answers is not in the document]
Comment (jpmcmu) [reply commons-hpcc/src/main/java/org/hpccsystems/commons/benchmarking/AveragedMetric.java:84]: Good catch

File: commons-hpcc/src/main/java/org/hpccsystems/commons/benchmarking/BenchmarkResult.java (hunk 2 above)
Comment (jpmcmu) [reply commons-hpcc/src/main/java/org/hpccsystems/commons/benchmarking/BenchmarkResult.java:133]: Good question. If we are reporting the average then probably not, outliers can skew the average significantly.

**Comment by jpmcmu** [comment]
@rpastrana  please review

File: commons-hpcc/src/main/java/org/hpccsystems/commons/benchmarking/BenchmarkResult.java [hunk 4]
```diff
@@ -156,11 +229,14 @@ public JSONArray toFlatJson(String[] selectedResults)
...
         if (selectedResults == null || selectedResults.length == 0)
         {
-            for (HashMap.Entry<String, AveragedMetric> entry : metrics.entrySet())
+            for (HashMap.Entry<String, MetricInfo> entry : metrics.entrySet())
             {
-                AveragedMetric metric = entry.getValue();
-                metric.discardOutliers();
-                
+                MetricInfo metric = entry.getValue();
+                if (metric.metric == null)
```
Comment (rpastrana) [inline commons-hpcc/src/main/java/org/hpccsystems/commons/benchmarking/BenchmarkResult.java:235]: if (metric || metric.metric == null) as above?
Also, mostly for aesthetics, but if we're going to make changes in this code, might as well refactor the metric var on lines 205,234 and 248 to metricinfo or currmetric

File: commons-hpcc/src/main/java/org/hpccsystems/commons/benchmarking/MetricAverageTransformer.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+
+/**
+ * Helper class that combines metrics with the same name into an AveragedMetric.
+ */
+public class MetricAverageTransformer
```
Comment (rpastrana) [inline commons-hpcc/src/main/java/org/hpccsystems/commons/benchmarking/MetricAverageTransformer.java:27]: is this meant to implement IMetricSetTransformer?

File: commons-hpcc/src/main/java/org/hpccsystems/commons/benchmarking/MetricAverageTransformer.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+{
+    public List<IMetric> transform(List<IMetric> metrics)
```
Comment (rpastrana) [inline commons-hpcc/src/main/java/org/hpccsystems/commons/benchmarking/MetricAverageTransformer.java:29]: is this meant to override IMetricSetTransformer.transform()?
If so, annotate with  @Override

File: commons-hpcc/src/main/java/org/hpccsystems/commons/benchmarking/MetricSumTransformer.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+
+/**
+ * Helper class that sums together metrics with the same name.
+ */
+public class MetricSumTransformer implements IMetricSetTransformer
+{
+    public List<IMetric> transform(List<IMetric> metrics)
```
Comment (rpastrana) [inline commons-hpcc/src/main/java/org/hpccsystems/commons/benchmarking/MetricSumTransformer.java:29]: @Override?

File: commons-hpcc/src/main/java/org/hpccsystems/commons/benchmarking/Units.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        public int scalePower = 0;
+        private Scale(String prefix, int scalePower)
+        {
+            this.scalePrefix = prefix;
+            this.scalePower = scalePower;
+        }
+    };
+    public Scale scale;
+    
+    public Units(Type type, Scale scale)
```
Comment (rpastrana) [inline commons-hpcc/src/main/java/org/hpccsystems/commons/benchmarking/Units.java:61]: does it make sense to provide a simpler ctr Units(Type type) and the Scale is defaulted to Scale.UNIT?

File: commons-hpcc/src/main/java/org/hpccsystems/commons/benchmarking/Units.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+    public String toString()
+    {
+        return scale.scalePrefix + type.typePostfix;
+    }
+
+
+    public static double calculateScaleConversion(Units.Scale from, Units.Scale to)
+    {
+        return Math.pow(10.0,(double)(from.scalePower-to.scalePower));
```
Comment (rpastrana) [inline commons-hpcc/src/main/java/org/hpccsystems/commons/benchmarking/Units.java:75]: does Math.pow provide the expected result if the exponent is negative?

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/RowServiceInputStream.java [hunk 10]
```diff
@@ -1116,69 +1137,24 @@ public long skip(long n) throws IOException
...
-        return numFetches;
-    }
-
-    public long getNumPartialBlockReads()
-    {
-        return numPartialBlockReads;
-    }
-
-    public long getNumBlockReads()
+    public List<IMetric> getMetrics()
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/RowServiceInputStream.java:1140]: @override?

File: commons-hpcc/src/main/java/org/hpccsystems/commons/benchmarking/Units.java (hunk 9 above)
Comment (jpmcmu) [reply commons-hpcc/src/main/java/org/hpccsystems/commons/benchmarking/Units.java:75]: Yep, Math.pow will handle the negative exponent and give us the correct result. IE bytes -> megabytes will give us 10e-6 that we would want

**Comment by jpmcmu** [comment]
@rpastrana changes made and squashed

File: commons-hpcc/src/main/java/org/hpccsystems/commons/benchmarking/BenchmarkResult.java [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+        {
+            return null;
+        }
+
+        return info.metric;
+    }
+
+    public void addMetric(IMetric metric)
+    {
+        MetricInfo info = metrics.get(metric.getName());
```
Comment (rpastrana) [inline commons-hpcc/src/main/java/org/hpccsystems/commons/benchmarking/BenchmarkResult.java:75]: could metric == null?

**Pull request merged**